### ISSUE TITLE
`BuiltinFunctionType` and `BuiltinMethodType` are different

### ISSUE THREAD

**Issue by Tetramad** [RustPython/RustPython#3100]
## Feature

`types.BuiltinFunctionType` must be equal to `types.BuiltinMethodType`.
In other word, types returned by builtin function and types returned by builtin method must be of the same type.

### In CPython
```py
>>> import types
>>> types.BuiltinFunctionType == types.BuiltinMethodType
True
>>> type(len) == type([].append)
True
```

### In RustPython
```py
>>>>> import types
>>>>> types.BuiltinFunctionType == types.BuiltinMethodType
False
>>>>> type(len) == type([].append)
False
```

## Python Documentation

[BuiltinFunctionType](https://docs.python.org/3/library/types.html#types.BuiltinFunctionType) and [BuiltinMethodType](https://docs.python.org/3/library/types.html#types.BuiltinMethodType) have same descriptions.
Also, commented at [source code](https://github.com/python/cpython/blob/050d1035957379d70e8601e6f5636637716a264b/Lib/types.py#L44).

**Comment by youknowone** [comment]
Roughly, adding `zelf: Option<PyObjectRef>` to `PyBuiltinFunction` and replacing `PyBoundMethod` to `PyBuiltinFunction` will solve it.

It looks like a good next-step issue for who understand how python binding works and finished a few good first issues.

**Comment by Snowapril** [comment]
This issue does not matter anymore.
```python
>>>>> import types
>>>>> types.BuiltinFunctionType == types.BuiltinMeth
odType
False
>>>>> type(len) == type([].append)
False
```

**Comment by moreal** [comment]
I'm working on this issue. I left this comment to avoid task conflicts.

**Comment by seungha-kim** [comment]
I'll take a look!

**Comment by youknowone** [comment]
Maybe, we can adapt the optimization idea but not about the exact types.
CPython internally uses `PyCMethod` but expose it as `PyCFunction`. I think that will be very tricky in rust.

**Comment by youknowone** [comment]
Partially resolved by #4873 
I decide not to strictly follow CPython implementation for this behavior. We expose BuiltinMethodType to user.
